components/threading/thread_viewer/thread_viewer.tsx

```diff
@@ -0,0 +1,564 @@
+                postsContainerHeight: containerHeight - (createContainerHeight + headerHeight),
+            });
+        }
+    }
+
+    public componentDidMount() {
+        this.scrollToBottom();
+        this.resizeRhsPostList();
+        window.addEventListener('resize', this.handleResize);
+
+        if (this.morePostsToFetch()) {
+            this.props.actions.getPostThread(this.props.selected.id, true);
+        }
+
+        if (this.props.isCollapsedThreadsEnabled) {
+            if (this.props.userThread == null) {
+                this.fetchThread();
+            } else {
+                this.markThreadRead();
+            }
+        }
+    }
+
+    public componentWillUnmount() {
+        window.removeEventListener('resize', this.handleResize);
+    }
+
+    public morePostsToFetch() {
+        return this.props.posts.length <= 1 || this.props.posts.length < (Utils.getRootPost(this.props.posts).reply_count + 1);
+    }
+
+    fetchThread() {
+        this.props.actions.getThread(
+            this.props.currentUserId,
+            this.props.currentTeamId,
+            this.props.selected.id,
+            true,
+        );
+    }
+
+    markThreadRead() {
+        if (
+            this.props.userThread &&
+            (
+                this.props.userThread.last_viewed_at < this.props.userThread.last_reply_at ||
+                this.props.userThread.unread_mentions ||
+                this.props.userThread.unread_replies
+            )
+        ) {
+            this.props.actions.updateThreadRead(
+                this.props.currentUserId,
+                this.props.currentTeamId,
+                this.props.selected.id,
+                Date.now(),
+            );
+        }
+    }
+
+    public componentDidUpdate(prevProps: Props) {
+        const prevPostsArray = prevProps.posts || [];
+        const curPostsArray = this.props.posts || [];
+
+        const reconnected = this.props.socketConnectionStatus && !prevProps.socketConnectionStatus;
+        const selectedChanged = this.props.selected.id !== prevProps.selected.id;
+
+        if (reconnected || selectedChanged) {
+            this.props.actions.getPostThread(this.props.selected.id);
+
+            if (
+                this.props.isCollapsedThreadsEnabled &&
+                this.props.userThread == null
+            ) {
+                this.fetchThread();
+            }
+        }
+
+        if (
+            this.props.isCollapsedThreadsEnabled &&
+            this.props.userThread?.id !== prevProps.userThread?.id
+        ) {
+            this.markThreadRead();
+        }
+
+        if (prevPostsArray.length >= curPostsArray.length) {
+            return;
+        }
+
+        const curLastPost = curPostsArray[0];
+
+        if (curLastPost.user_id === this.props.currentUserId || this.state.userScrolledToBottom) {
+            this.scrollToBottom();
+        }
+    }
+
+    public shouldComponentUpdate(nextProps: Props, nextState: State) {
```

Comment:
Are we certain `shouldComponentUpdate` is required here? Usually this is a performance improvement, but only needed if we've measured one. And if we have to compute the equality of lots of things anyway, it may be a wash. 

On the flip side, if we're using this to work around a problem of updating when we shouldn't, let's identify and tackle the root problem there.

Reply:
This was already here, and I'm not 100% sure of the history, but the only change I added was was for if the `userThread` was different.

I agree though that it may not be needed or the root problem should be addressed in a more appropriate way, though I do think any significant refactoring should be done separately.

`ThreadViewer` is really just the guts of `RHSThread` extracted to be reusable.